### PR TITLE
Enhancement: Optional Qualtrics Credentials

### LINKED ISSUE #53: Multiple Valid Credentials must be provided to use Token ATM
At the moment, valid credentials for both Canvas and Qualtrics must be provided to make it past the login page.

Mandatory Canvas credentials seem reasonable. 
But not all courses need/use Qualtrics.

For more info, see this pull request: https://github.com/UCI-IN4MATX-191-Token-ATM/token-atm-spa/pull/52#issuecomment-1631655991

### PR DESCRIPTION
**Pull request by yongkanm**
### Description

This pull request makes Token ATM no longer forcefully requires Qualtrics credentials. The following features are provided to enable that:

1.  If the user have saved credentials, when a user log in, the user could now choose to unlock the credentials with a password but not directly log in. This feature allows the user to make changes to saved credentials. Please note that changes will not be automatically saved unless the user logs in with the “Save credentials” checkbox checked and provides a new password.
2.  Field to provide Qualtrics credentials on the login page are removed. Instead, a new section named “Optional Credentials” is added to the login page. This section contains multiple buttons to open modals used for configuring credentials for optional third-party services. Currently the only third-party service is Qualtrics. Please note that any modification done in the modal will only have effect in the current session unless changes are saved as mentioned above.
      
    For each button, a circle is shown at the right side of the service name. Whether the circle is filled indicates whether the user have provided credential for this service.
      
    If the user have provided credential for a service, the user will be able to delete provided credential in the modal for this service.
      
    When clicking “Configure” button in the modal to save provided credential, a validation of the provided credential will be performed.
3.  When logging in without providing Qualtrics credentials, token options with the type of `earn-by-survey` will be marked as “Missing Credentials” in the “Token Options” page. A confirmation modal will show up to ask if the user want to provide Qualtrics credential when the user attempts to view or edit these token options. The user could still delete these token options using the drop-down menu.
4.  When logging in without providing Qualtrics credentials, if there are token options with the type of `earn-by-survey` configured for this course, a modal will be shown to warn the user that there are missing credentials if the user attempts to start request processing. From this modal, the user could choose to cancel request processing, go back to the login screen to provide missing credentials, or start the request processing anyway.

### Testing Note

1.  This pull request introduces a non-backward-compatible change to the format of saved credentials. Therefore, it is suggested to use a separated environment (e.g., a different port than 4200) to test whether saving/unlocking credentials still work as expected. Also, please test if the migration from the old format to the new format works as expected by saving credentials in the old version and then test whether they work in the new version.
2.  This pull request makes changes to the startup process of the Angular application. To be specific, `AppModule` is now imported asynhrhouously. This is to allow classes with decorators that inject services to Angular DI being loaded before `AppModule` getting loaded.
      
    This change will leads to the change of build output. Please test if the build process still works as expected by building the project and then hosting the build output to see if it still works.
3.  Please test if features mentioned above work as expected.

### CODE REVIEW

**Review by epsilonError — CHANGES_REQUESTED** [review]
Code and implementation looks good.

Came across one aspect that must be addressed ([See this note](https://github.com/UCI-IN4MATX-191-Token-ATM/token-atm-spa/pull/103#issuecomment-1861769378)).

If you have time the other comments can be addressed as well, but aren't required for future approval.

**Comment by yongkanm** [comment]
> **Note**: New Optional Credential Modal removes tool-tip information about input fields.
> 
> But the help button does link to the same information.

Yes. That is because currently FormField does not support inline help message. It is planned to be supported at the next refactoring of FormField.

**Comment by yongkanm** [comment]
> **Note**: Users may accidentally delete encrypted saved credentials. Delete button, on Load Saved Credentials modal, doesn't ask for confirmation before wiping Local Storage.
> 
> **Should be Fixed.**

Thanks for pointing this out! This issue also exists in the old versions. I will fix it within this pull request.

**Comment by yongkanm** [comment]
> **Note**: Users are prevented from viewing a Token Option Configuration, but they don't actually need active optional credentials to view. They need credentials to edit.
> 
> Shouldn't we prevent editing, not viewing as well?
> 
> This change does require some functionality change in Token Option Configuration modals though:
> 
> * Enumeration calls for selection and multi-selection fields need to know not to make the outbound call and handle just displaying the provided configuration.
> * Others?

Agreed. We should let the user view what they've configured in any case. For editing token option, a warning modal will show up to notify the user that credentials are missing. The user could choose to stay in the viewing mode, go back to the login screen to provide credentials, or enter the editing mode anyway. If the user choose to edit anyway, their change could never be saved because the validation will always fail due to the missing credentials.

**Comment by yongkanm** [comment]
> **Note**: Users may end up waiting an interminable amount of time before being notified about network retries.
> 
> Can we provide a toast message/notification that network retries are/will be attempted?
> 
> To test: Disconnect from the internet, and attempt to login with Qualtrics credentials.

That's a great idea! I will implement it in this pull request.

**Comment by yongkanm** [comment]
> **Note**: Users can start adding a token option for `earn-by-survey` even if they don't have valid credentials.
> 
> The validator will never pass, because they don't have access to Qualtrics, but the Create Token Option modal does appear.
> 
> (P.S. The validator does display the most relevant error though: `Error: Credentials for Qualtrics are invalid`.)
> 
> If we're preventing viewing an already created option, should we also prevent viewing the Token Option creation modal?
> 
> I'm open to users being able to see both, but then the "Create" or "Edit" buttons should show the 'go back to login' modal (or be disabled) if the needed credential isn't present.

I originally planned to disable the creation of token option that requires missing credentials, but based on our discussion about viewing token option with missing credentials, I decide to create a warning modal for the user after they click "Next Step" in the modal used for selecting the type of token option to be created. This modal have the same format and functionalities but different message as the warning modal shown when the user attempts to edit a token option that requires missing credentials (see details in [this comment](https://github.com/UCI-IN4MATX-191-Token-ATM/token-atm-spa/pull/103#issuecomment-1863747411)).

**Comment by yongkanm** [comment]
Changes mentioned in the discussion above are now available. In addition, there is a new change to the login page: when the user clicks the "Login" button, fields and buttons are disabled to prevent the user from modifying them during the validation.

**Review by epsilonError — DISMISSED** [review]
Enhancement is approved. All necessary to fix comments have been addressed.

A few notes are left unresolved and will be completed in future pull requests.

File: src/app/components/create-token-option-modal/create-token-option-modal.component.html [hunk 1]
```diff
@@ -14,7 +20,16 @@ <h4 class="modal-title pull-left">Create Token Option</h4>
...
-        <button type="button" class="btn btn-primary" (click)="onNextStep()">Next Step</button>
+        <button
+            type="button"
+            class="btn btn-outline-secondary me-4"
+            (click)="modalRef?.hide()"
+            [disabled]="isProcessing"
+        >
+            Cancel
+        </button>
+        <button type="button" class="btn btn-primary" (click)="onNextStep()" [disabled]="isProcessing">
```
[the remark this answers is not in the document]
Comment (yongkanm) [reply src/app/components/create-token-option-modal/create-token-option-modal.component.html:31]: Angular event binding does not support awaiting for asynchorous calls, so the event handler will invoke the function and returns immediately.

**Comment by yongkanm** [comment]
> > > **Note**: Users may end up waiting an interminable amount of time before being notified about network retries.
> > > Can we provide a toast message/notification that network retries are/will be attempted?
> > > To test: Disconnect from the internet, and attempt to login with Qualtrics credentials.
> > 
> > 
> > That's a great idea! I will implement it in this pull request.
> 
> Implemented! Though on my machine the message is quite small and disappears before I can finish reading it. This is probably something to improve in a future feature request. If it's easy to extend the time then feel free to update, but we can leave it for later.

The duration is short because I align it with the exponential backoff waiting time. This is indeed not user-friendly. I will make the snackbar continually exist during all retries. If the user dismisses it, while another retry occurs, it will be shown again.

**Comment by yongkanm** [comment]
> > This also involves a change to the existing FormField(s) to ensure that user's existing configuration could be shown even if errors occurred.
> 
> I'm not sure how to test this. But the code change looks good.
> 
> I managed to try testing by pausing the canvas docker container. At least for network errors, the select and multi-select field don't show the already configured fields since the input field never finishes its "loading..." phase.
> 
> But fixing this should be done in another feature pull request.

That is interesting because behavior on my side is different. First, I've tested it by intentionally throw an error from the `optionFactory` pass into `MultipleSelectionField` or `SingleSelectionField`. An example is to throw an error from line `197-200` in `token-option-field-component-factory.ts`. The behavior is throwing an error for failing to set options, but shows all previously selectecd options as invalid options (because they cannot be validated).

I also tested a network error by turning off my Ethernet connection. Behavior on my side is that the snackbar retry message is shown, and after several retries, the behavior is the same as intentionally throwing an error from `optionFactory`.

I suspect the difference might be from how we disconnect from the network. Can you include more details here?

**Comment by yongkanm** [comment]
> > > > This also involves a change to the existing FormField(s) to ensure that user's existing configuration could be shown even if errors occurred.
> > > 
> > > 
> > > I'm not sure how to test this. But the code change looks good.
> > > I managed to try testing by pausing the canvas docker container. At least for network errors, the select and multi-select field don't show the already configured fields since the input field never finishes its "loading..." phase.
> > > But fixing this should be done in another feature pull request.
> > 
> > 
> > That is interesting because behavior on my side is different. First, I've tested it by intentionally throw an error from the `optionFactory` pass into `MultipleSelectionField` or `SingleSelectionField`. An example is to throw an error from line `197-200` in `token-option-field-component-factory.ts`. The behavior is throwing an error for failing to set options, but shows all previously selectecd options as invalid options (because they cannot be validated).
> > I also tested a network error by turning off my Ethernet connection. Behavior on my side is that the snackbar retry message is shown, and after several retries, the behavior is the same as intentionally throwing an error from `optionFactory`.
> > I suspect the difference might be from how we disconnect from the network. Can you include more details here?
> 
> I used `docker pause Canvas-Test` once I'd opened the Token Option page. And the selection fields never populated to even say they were invalid.
> 
> And did also try with the instructure canvas and I just flipped a switch on my hypervisor to cut off all traffic (but with that the network cards and interfaces still believe they are all active, the hypervisor just blocks all traffic). I don't remember anything different happening there. But the retry snackbars did appear.

If the retry snackbar appears, then the error should show up after several retries (approx 10 seconds). `docker pause` might have entirely different behavior for network than a network disconnection. My suspection is that what plays a role here is the TCP timeout, which could last for several minutes. Maybe we should have an internal timeout mechanism implemented, but this will require changed to `token-atm-app`.

**Review by epsilonError — APPROVED** [review]
Network retries snackbars now persist.
And (multi-)selection inputs allow previewing the existing configuration data, if an error occurs.

Network errors/retries may postpone the viewability of the existing configuration, but this isn't an enhancement blocking problem.

All tests are passed, and all comments/change-requests have been addressed.

**Comment by yongkanm** [comment]
> Regarding the snack bar for network retry notification,
> 
> > If the user dismisses it, while another retry occurs, it will be shown again.
> 
> I assumed this meant there was a backlog of snackbar notifications. E.g. if two timeout retries had happened with any snackbars being dismissed, the user would have to click the dismiss button twice so remove the snackbar from the screen.
> 
> Currently one dismiss click clears an entire backlog in one go (e.g. 1 dismiss clears all pending snackbars in one go). I think this is a good implementation, but I'm not sure if this was what you were describing above.

Sorry for not make this clear, the effect you observed is expected. `ExponentialBackoffExecutor.execute()` now manually maintain the lifecycle of a snackbar. When a retry occurs, a snackbar will be created if either the old snackbar is dismissed by the user or this is the first retry. The snackbar is programmatically dismissed when `ExponentialBackoffExecutor.execute()` returns (by either throwing an error or return a value). There won't be multiple snackbar for a single request.

**Comment by yongkanm** [comment]
> > `docker pause` might have entirely different behavior for network than a network disconnection.
> 
> This is true. It seems like it becomes a black hole at the moment. Maybe a docker port/interface keeps running even though the container is frozen.
> 
> But my hypervisor network cutoff switch does result in retries and snackbar notifications. So I must have been misremembering that, or I hadn't pulled the update for persistent snackbar notifications when I tried it.
> 
> The docker example makes me think of making a more robust network error detection/notification system. But that is more work than we need to take on.



> > That is interesting because behavior on my side is different. First, I've tested it by intentionally throw an error from the `optionFactory` pass into `MultipleSelectionField` or `SingleSelectionField`. An example is to throw an error from line `197-200` in `token-option-field-component-factory.ts`. The behavior is throwing an error for failing to set options, but shows all previously selectecd options as invalid options (because they cannot be validated).
> > I also tested a network error by turning off my Ethernet connection. Behavior on my side is that the snackbar retry message is shown, and after several retries, the behavior is the same as intentionally throwing an error from `optionFactory`.
> 
> I now have the same results. On thrown error, the field is populated with the current configured value and is red because it cannot be validated.
> 
> On network disconnect, I might be having a different result from you. With my network access turned off, the selection box stays empty while network retries are attempted and the field says it is loading. This can mean it can be quite a long time before the already configured field is shown. Is an empty selection box that say loading below it what you get while ethernet disconnected?
> 
> If it is possible to force the showing/preview of the current selection (even though it cannot be validated at the moment), and then empty / fill the selection list at the last possible moment?
> 
> But I could see a case being made that not showing the unvalidated red, will be confusing if that appears right way before the notifications about network retries.
> 
> As I mentioned before, this is probably a fix to fit in a future feature/enhancement update. And the improvements so far are already quite good.

Agreed. We indeed need more effort to figure out a better system to handle network issues. Also, `FormField` will possibly need refactoring for better usability. I'll merge this pull request for now since these possible improvements are not sufficiently related to the support of optional Qualtrics credential.

**Pull request merged**